### PR TITLE
Feature GitHub webhooks

### PR DESCRIPTION
**Pull request by YutongZhuu**
Installed Express.js for subscribing webhook from gitHub, fixed the absolute url issue, the feature is not fully working; however, people working on other feature also might need a web server, so I decided to create a PR now.

### CODE REVIEW

**Comment by carlxw** [comment]
Why are we combining a ExpressJS backend and a discord bot in the same file? I'm pretty sure we don't want that down the road here.

**Comment by YutongZhuu** [comment]
> Why are we combining a ExpressJS backend and a discord bot in the same file? I'm pretty sure we don't want that down the road here.

In my understanding, we need a http server to listen to webhooks. Correct me if I am wrong.

**Comment by carlxw** [comment]
If so, it's better to create the RESTful API in it's separate folder at least, rather than throwing everything inside `index.js`. It's probably even better for the server to have it's own Github repository.

**Comment by YutongZhuu** [comment]
> Why are we combining a ExpressJS backend and a discord bot in the same file? I'm pretty sure we don't want that down the road here.

I am not sure if you are asking about the usage of Express or the problem with placing them in the same file. I think we need to start both discord and express at the same time, so it is convenient to put them in the index.js file.

**Comment by YutongZhuu** [comment]
> If so, it's better to create the RESTful API in it's separate folder at least, rather than throwing everything inside `index.js`. It's probably even better for the server to have it's own Github repository.

Yes, sure, I can do that, but I dont think placing the express server in another git repo is doable

**Comment by carlxw** [comment]
@YutongZhuu I don't see why having the server in another repo has any hardships

@sherwinchiu What are your thoughts on this

**Comment by sherwinchiu** [comment]
I don't really see the point of having this in another repo - it's a part of the Discord bot even if we're hosting another express server for ChatOps. Both the server and the discord bot must be on for chat-ops to work, so it makes sense that they both work together. I think it makes even more sense to separate the main discord bot and express.js server in two different files because of this as well. It provides separation while keeping them together

**Comment by sherwinchiu** [comment]
Index.js should just be for making the discord bot work, express is a separate thing so it belongs in a separate file imo too

**Comment by sherwinchiu** [comment]
since this is a pretty small team, and the bot won't have any crazy features, we should just keep one repo :)

**Comment by carlxw** [comment]
Nice. Thanks for the insight Sherwin.

File: index.js [hunk 1]
```diff
@@ -14,6 +23,13 @@ client.once(Events.ClientReady, readyClient => {
 	console.log(`Ready! Logged in as ${readyClient.user.tag}`);
 });
 
+client.on("messageCreate", msg => {
```
Comment (carlxw) [inline index.js]: We are using slash commands to interact with the bot. There's no need to create an event for `messageCreate`

Comment (YutongZhuu) [reply index.js]: right

Comment (YutongZhuu) [reply index.js]: That was for my testing purpose, I forgot to delete it :)

**Comment by YutongZhuu** [comment]
@carlxw  I am not sure how can I refactor the code. since the express app will need the 'client' object. If I require the client the object from another file, it will re-run everything in the index.js file, which does not seem to be a good practice. So, I think what doable is that we should put the client object config in another file, and put the express server in the index.js file.

**Comment by carlxw** [comment]
> So, I think what doable is that we should put the client object config in another file, and put the express server in the index.js file.

No. index.js is reserved for the the Discord bot only. If you need a particular object from `index.js`, you can export that particular object (in this case, client) and import it into the ExpressJS main file.

I believe the code line you need is [`module.exports`](https://www.sitepoint.com/understanding-module-exports-exports-node-js/).

I don't understand with what you mean by "re-run everything from the index file", but I'd imagine that it should be fine as long as ExpressJS and DiscordJS are kept separate

**Comment by YutongZhuu** [comment]
What I did was create another client instance and login in the express server file, I cannot find any better solution. (exporting from index file did not work too). So the drawback is that we need to run two commands in order to start our app, but it can be fixed by configuring npm I believe. I will push my changes.

**Comment by YutongZhuu** [comment]
And you are right, we could potentially create another repo for the server, but it will definitely a bit too hard for us to manage.

**Comment by carlxw** [comment]
I believe that is the expectation. One command to start the bot, another command (in new terminal) to start the Express app. Regarding the client exporting, I now see why that probably caused issues.

**Comment by sherwinchiu** [comment]
I'm sure there's a way to get both objects in two different files. Have a shared object instance for example that can be shared and passed through, or maybe make a separate file for the shared object.

**Review by sherwinchiu — APPROVED** [review]
i think this is fine for now to get webhooks working. Can you get it working with the discord bot on our server first with a new channel? ill gave you admin perms

**Pull request merged**